pkg/skaffold/docker/default_repo.go

```diff
@@ -49,6 +58,38 @@ func SubstituteDefaultRepoIntoImage(defaultRepo string, image string) (string, e
 	return replaced, nil
 }
 
+func substituteDefaultRepoIntoImageNew(defaultRepo string, image string) (string, error) {
+	if defaultRepo == "" {
+		return image, nil
+	}
+
+	parsed, err := ParseReference(image)
+	if err != nil {
+		return "", err
+	}
+
+	// replace registry in parsed name
+	var replaced string
+	reg, image := splitImage(parsed.BaseName)
+	defaultRegistry := registry.GetRegistry(defaultRepo)
+	newReg := reg.Update(defaultRegistry)
+	if newReg.Type() == reg.Type() {
+		replaced = newReg.Name() + "/" + image
+	} else {
+		replaced = newReg.Name() + "/" + escapeRegex.ReplaceAllString(parsed.BaseName, "_")
+	}
```

Comment:
I found this section really hard to figure out what was going on.  It all seems to hinges on Registry.Update() returning its argument (`defaultRegistry`) unless they're both `GCRRegistry`.

Reply:
or both are `GenericRegistry`. I added some more test cases which will help clarify this more.